rfcs/rfc-007-dynamic-program-dispatch.md

```diff
@@ -0,0 +1,183 @@
+ }
+
+ impl Account {
+@@ -19,6 +25,8 @@ impl Account {
+             tokens,
+             userdata: vec![0u8; space],
+             program_id,
++            false,
++            None,
+         }
+     }
+ }
+```
+
+
+The executable flag may only be set by the `program_id` the account belongs to.
+Once the executable flag is set:
+1. The bank enforces the account userdata read-only even by `program_id`
+2. The executable flag cannot be unset
+3. The account may be used as input to `SystemProgram::Spawn` to forge a new `program_id`
+
+The `loader_program_id` field is set by `SystemProgram::Spawn` and is used by
+bank to locate the required loader during transaction dispatch.
+
+### Common Loader Interface
+
+```rust
+//! Instructions that must be implemented by all Loaders
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum LoaderInstruction {
+    /// Write program data into an account
+    ///
+    /// * key[0] - the account to write into.
+    ///
+    /// The transaction must be signed by key[0]
+    Write { offset: u32, prog: Vec<u8> },
+
+    /// Prepare an account loaded with program data for execution.
+    /// The exact preparation steps is loader specific but on success the loader must set the executable
+    /// bit of the Account
+    ///
+    /// * key[0] - the account to prepare for execution
+    ///
+    /// The transaction must be signed by key[0]
+    Prepare,
+}
+```
+
+`LoaderInstruction` is only used for transactions where the account at
+`Transaction::key[N]` is not executable.
+
+The loader also includes the *program interpreter* within it.  On transactions
+where the `Transaction::key[N]` account is executable and has a
+`loader_program_id`, the loader instead invokes the program in
+`Transaction::key[N]` passing along the transaction userdata and only
+`Transaction::key[1..N-1]`.
+
+
+### SystemProgram additions
+
+```rust
+    /// Transforms `executable_account_id` into a valid `program_id` for future transactions.
+    /// This transaction must be signed by `executable_account_id`.
+    ///
+    /// executable_account_id: An account with its `executable` flag set and `loader_program_id.is_none()`
+    ///
+    /// On success:
+    /// * The current `program_id` of `executable_account_id` will be saved in `loader_program_id`
+    /// * The `program_id` of `executable_account_id` will be set to `executable_account_id`
+    ///
+    SystemProgram::Spawn{
+        executable_account_id: Pubkey,
+    }
+```
+
+### Native Loader
+
+NativeLoader implements the `LoaderInstruction` interface as follows:
+
+1. User writes the name of the native module (`solua`, `move_funds`, `noop`) into the account userdata using `LoaderInstruction::Write`
+2. On `LoaderInstruction::Prepare`, NativeLoader:
+    1. loads the corresponding native module into memory
+    2. modifies the account userdata to contain a 0-based index into its internal vector of loaded shared objects
```

Comment:
I intentionally didn't think harder here.  It might not even be a vector of shared objects.  The point is that there's some information in a NativeLoader account userdata that allows NativeLoader to locate the matching binary.  It could even keep the original name value written via `LoaderInstruction::Write`, actually maybe it should :)